### DISCUSSION TITLE
How is geograpy3 determining locations from text?

### DISCUSSION THREAD

**Discussion by Dan-Price-Intrepid** [somnathrakshit/geograpy3#61] [Q&A] [answered]
I have an example text string that I'm running as follows:

`places = geograpy.get_place_context(text="Egypt again, Russia, India and Uzbekistan.  ")`

The result is:
['India', 'Egypt', 'Russia', 'Uzbekistan', 'United States of America', 'Australia', 'Moldova']

When I expected just the countries as named in the text (e.g. Egypt, Russia, India and Uzbekistan).  Why is this happening?

**Comment by WolfgangFahl** [comment, marked as answer]
There might be places in the US, Australia and Moldova that are named "Egypt", "Russia", "India" or "Uzbekistan".
E.g.
```sql
select * from city_labels where label in ("Russia")
```
will return 
* https://www.wikidata.org/wiki/Q20657737 - Russia / township in Morris County, New Jersey
* https://www.wikidata.org/wiki/Q3708651 - Russia / town in New York
* https://www.wikidata.org/wiki/Q2477732 - Russia / town in Ohio

You might want to be more specific and use the locator interface instead and explicitly look for the country information. The get_place_context is outdated and doesn't have this capability.

OSMPythonTools.nominatim is also an excellent option. 
see https://github.com/WolfgangFahl/ConferenceCorpus/blob/main/corpus/location.py (Excerpt of the code below ...)
```python
from geograpy.locator import LocationContext
from OSMPythonTools.nominatim import Nominatim 
import os
import logging

class LocationLookup:

def __init__(self):
        '''
        Constructor
        '''
        cacheRootDir="cache"
        cacheDir=f"{cacheRootDir}/.nominatim"
        if not os.path.exists(cacheDir):
            os.makedirs(cacheDir)
            
        self.nominatim = Nominatim(cacheDir=cacheDir)
        logging.getLogger('OSMPythonTools').setLevel(logging.ERROR)

 def lookupNominatim(self,locationText:str):
        location=None
        nresult=self.nominatim.query(locationText,params={"extratags":"1"})
        nlod=nresult._json
        if len(nlod)>0:
            nrecord=nlod[0]
            if "extratags" in nrecord:
                extratags=nrecord["extratags"]
                if "wikidata" in extratags:
                    wikidataID=extratags["wikidata"]
                    location=self.getCityByWikiDataId(wikidataID)
        return location
```

**Reply by Dan-Price-Intrepid** [reply]
Thanks very much! 

I put together the following little code snippet and it seemed to do the trick:

```
from geograpy import places
from geograpy import extraction

e = extraction.Extractor(text="Egypt again, Russia, India and Uzbekistan. ")
e.find_geoEntities()
pc = places.PlaceContext(e.places)

pc.set_countries()
print (pc.countries )

```
